_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+
+#### OCFL
+
+Implementation notes for OCFL recommend things such as pair-trees for storing
+objects in the object root. I'm going to treat that as an implementation detail
+and pretend we're using a file system that can handle an infinite number of
+resources in the same directory.
+
+1. Create a base directory named `<book-id>`
+1. Create a file named `0=ocfl_object_1.0` with the appropriate contents.
+1. Create a v1 directory
+1. Copy all content files and metadata files into `<book-id>/v1/content`, separated
+   by ID. So, for instance, `<page1>` would be in `<book-id>/v1/content/<page1>`,
+   and have two files: `<page1>.json` and
+   `<page1>-<file-name>.<file-extension>`.
+1. Create an `inventory.json` in `v1`, as per the specification, with SHA512s of
+   every file in `content` as the keys of the `manifest` property.
+1. Create `inventory.json.sha512` in the `v1` directory.
+1. Copy `inventory.json` and `inventory.json.sha512` to the `<book-id>`
+   directory.
+
+**Pros**:
+
+1. We now have an OCFL repository which is theoretically readable by future OCFL
+   browsers, and will be understood by partner institutions.
+1. This structure just needs a file system to store itself and support
+   versioning, there's no need for an object store's feature-set.
+1. We can de-duplicate as part of generating the `inventory.json` if we'd like
+   to, using up less space than the other options.
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the inventory will be
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved as a new version of the bag, updating the
+      `inventory.json` and incrementing the version counter.
+   5. Initial preservation ends. Updates inventory, but has out-of-date checksum
+      information.
+   6. However, since versions are numbered we could theoretically check to see
+      before we start writing things if files exist to reduce the time we need
+      to lock.
+
+#### Google Cloud Storage
+
+A small note about GCS - I'm going to call paths "directories", but in GCS it's
+  all just naming semantics. `foo/bar/txt.yml` isn't a file in the directory
+  `bar`, it's a file that has a name which happens to have slashes in it. The
+  GCS tools just visualize them as directories to be nice.
+
```

Comment:
This is also how OCFL does things. Files are their hashes with a name so it's easier for humans.

Reply:
Yeah, but OCFL isn't a platform so I can't say a similar thing right? Like, if I delete foo/bar/txt.yml, then foo/bar is gone. The spec says I should just ignore foo/bar for OCFL, but can I say it won't be there?

Reply:
if you want to keep foo/bar you should create foo/bar/.keep to future proof storage migrations (because another system might not keep foo/bar)

Reply:
I'll add something to this section.

Reply:
Pushed up a small paragraph. Look okay @rosy1280?